File: src/mvmctl/core/cloudinit/_provisioner.py
"""Cloud-init provisioning for VM creation."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from mvmctl.constants import CONST_DIR_PERMS_CACHE
from mvmctl.exceptions import (
    CloudInitIsoModeError,
    CloudInitNetModeError,
)
from mvmctl.models import (
    CloudInitMode,
    FirewallChain,
    FirewallPort,
    FirewallProtocol,
    FirewallRule,
    FirewallRuleType,
    FirewallTable,
    FirewallTarget,
    FirewallWildcard,
    NetworkItem,
)
from mvmctl.services.nocloud_server.manager import NoCloudNetServerManager


@dataclass
class CloudInitProvisionConfig:
    """Configuration for cloud-init provisioning."""

    mode: CloudInitMode

    vm_name: str
    vm_id: str
    vm_dir: Path
    cloud_init_dir: Path

    guest_ip: str
    user: str
    tap_name: str
    network: NetworkItem
    network_prefix_len: int
    skip_network_config: bool

    ssh_pubkeys: list[str]

    # Resolved from defaults.cloudinit
    cloud_init_iso_name: str
    nocloud_port_range_start: int
    nocloud_port_range_end: int
    nocloud_max_port_retries: int

    custom_user_data_path: Path | None = None

    nocloud_net_port: int | None = None
    cloud_init_iso_path: Path | None = None
    keep_cloud_init_iso: bool = False


@dataclass
class CloudInitProvisionResult:
    """Result of cloud-init provisioning."""

    mode: CloudInitMode
    iso_path: Path | None = None
    nocloud_url: str | None = None
    nocloud_port: int = 0
    nocloud_pid: int | None = None
    nocloud_net_manager: NoCloudNetServerManager | None = None
    nocloud_net_rules: list[FirewallRule] = field(default_factory=list)


class CloudInitProvisioner:
    """Handle all cloud-init modes cleanly."""

    _config: CloudInitProvisionConfig

    def __init__(self, config: CloudInitProvisionConfig) -> None:
        """Initialize the CloudInitProvisioner."""
        from mvmctl.core.cloudinit._manager import CloudInitManager

        self._config = config
        self._manager = CloudInitManager(config)

    def provision(self) -> CloudInitProvisionResult:
        """
        Provision cloud-init based on configuration.

        Args:
            config: CloudInitProvisionConfig containing all provisioning parameters.

        Returns:
            CloudInitProvisionResult with what was created (iso path, nocloud url, etc.).

        """
        if self._config.mode == CloudInitMode.OFF:
            return self._provision_off()

        # Prepare the cloud-init configs
        self._config.cloud_init_dir.mkdir(
            mode=CONST_DIR_PERMS_CACHE, exist_ok=True
        )

        self._manager.write_config_files()

        result = None

        if self._config.mode == CloudInitMode.NET:
            result = self._provision_net()
        elif self._config.mode == CloudInitMode.ISO:
            result = self._provision_iso()
        else:
            result = self._provision_inject()

        return result

    def _provision_off(self) -> CloudInitProvisionResult:
        """Provision with cloud-init disabled."""
        return CloudInitProvisionResult(mode=CloudInitMode.OFF)

    def _provision_net(self) -> CloudInitProvisionResult:
        """Provision using nocloud-net mode with HTTP server."""

        from mvmctl.services.nocloud_server.manager import (
            NoCloudNetServerManager,
        )

        try:
            net_manager = NoCloudNetServerManager(
                id=self._config.vm_id,
                path=self._config.vm_dir,
                name=self._config.vm_name,
                ipv4_gateway=self._config.network.ipv4_gateway,
                port=self._config.nocloud_net_port
                if self._config.nocloud_net_port is not None
                else 0,  # Zero triggers auto-allocation from port range
                port_range_start=self._config.nocloud_port_range_start,
                port_range_end=self._config.nocloud_port_range_end,
                max_port_retries=self._config.nocloud_max_port_retries,
            )
            url, port, pid = net_manager.start()

            from mvmctl.core._shared import Database, FirewallTracker

            tracker = FirewallTracker(Database())
            tracker.ensure_chain(
                FirewallChain.MVM_NOCLOUDNET_INPUT, auto_jump_from="INPUT"
            )

            nocloud_net_in_rule = FirewallRule(
                table_name=FirewallTable.FILTER,
                chain_name=FirewallChain.MVM_NOCLOUDNET_INPUT,
                rule_type=FirewallRuleType.NOCLOUDNET_INPUT,
                target=FirewallTarget.ACCEPT,
                network_id=self._config.network.id,
                protocol=FirewallProtocol.TCP,
                source=self._config.guest_ip,
                destination=self._config.network.ipv4_gateway,
                in_interface=self._config.tap_name,
                out_interface=FirewallWildcard.ANY_INTERFACE,
                sport=FirewallPort.ANY,
                dport=port,
                network_name=self._config.network.name,
                comment_tag=f"# nocloudnet:{self._config.vm_name}:{port}",
                is_active=True,
            )
            tracker.ensure_rule(nocloud_net_in_rule)

            return CloudInitProvisionResult(
                mode=CloudInitMode.NET,
                nocloud_url=url,
                nocloud_port=port,
                nocloud_pid=pid,
                nocloud_net_manager=net_manager,
                nocloud_net_rules=[nocloud_net_in_rule],
            )
        except Exception as exc:
            raise CloudInitNetModeError(
                f"Nocloud-net provisioning failed: {exc}"
            ) from exc

    def _provision_iso(self) -> CloudInitProvisionResult:
        """Provision using ISO mode with cloud-init ISO image."""

        if self._config.cloud_init_iso_path is not None:
            if not self._config.cloud_init_iso_path.exists():
                raise CloudInitIsoModeError(
                    f"Custom cloud-init ISO not found: {self._config.cloud_init_iso_path}"
                )
            return CloudInitProvisionResult(
                mode=CloudInitMode.ISO,
                iso_path=self._config.cloud_init_iso_path,
            )

        iso_path = self._config.vm_dir / self._config.cloud_init_iso_name
        try:
            self._manager.create_seed_iso(self._config.cloud_init_dir, iso_path)
        except Exception as exc:
            raise CloudInitIsoModeError(
                f"Failed to create cloud-init ISO: {exc}"
            ) from exc

        return CloudInitProvisionResult(
            mode=CloudInitMode.ISO, iso_path=iso_path
        )

    def _provision_inject(self) -> CloudInitProvisionResult:
        """Provision using inject mode — config files already written by provision()."""
        return CloudInitProvisionResult(mode=CloudInitMode.INJECT)

File: internal/core/cloudinit/provisioner.go
package cloudinit

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"time"

	"mvmctl/internal/infra"
	"mvmctl/internal/lib/firewall"
	"mvmctl/internal/lib/model"
	nocloudnetsvc "mvmctl/internal/service/nocloudnet"
	"mvmctl/pkg/errs"
)

// Default auto-kill timeout for spawned nocloud-net server subprocess.
// Matches Python's _KILL_NO_CLOUD_SERVER_AFTER.
const defaultNocloudKillAfter = 5 * time.Minute

// Provisioner handles all cloud-init provisioning modes.
// Matches Python's CloudInitProvisioner.
type Provisioner struct {
	config          *Config
	manager         *Manager
	firewallTracker *firewall.FirewallTracker
}

// NewProvisioner creates a new CloudInitProvisioner.
// The tracker parameter is a pre-configured firewall tracker. If nil,
// firewall operations in NET mode are skipped.
func NewProvisioner(config *Config, tracker *firewall.FirewallTracker) *Provisioner {
	return &Provisioner{
		config:          config,
		manager:         NewManager(config),
		firewallTracker: tracker,
	}
}

// Provision performs cloud-init provisioning based on the configured mode.
// Matches Python's provision().
func (p *Provisioner) Provision(ctx context.Context) (*model.CloudInitResult, error) {
	if p.config.Mode == model.CloudInitModeOFF {
		return p.provisionOff(ctx), nil
	}

	// Prepare the cloud-init config directory — Python uses CONST_DIR_PERMS_CACHE = 0o700
	if err := os.MkdirAll(p.config.CloudInitDir, 0700); err != nil {
		return nil, errs.New(errs.CodeCloudInitProvisionFailed,
			fmt.Sprintf("create cloud-init dir: %s", err))
	}

	// Generate config files
	if err := p.manager.Generate(ctx); err != nil {
		return nil, err
	}

	switch p.config.Mode {
	case model.CloudInitModeNET:
		return p.provisionNet(ctx)
	case model.CloudInitModeISO:
		return p.provisionISO(ctx)
	case model.CloudInitModeINJECT:
		return p.provisionInject(ctx)
	default:
		return nil, fmt.Errorf("unknown cloud-init mode: %s", p.config.Mode)
	}
}

// provisionOff handles OFF mode — cloud-init disabled.
// Matches Python's _provision_off().
// Python: CloudInitmodel.CloudInitResult(mode=CloudInitMode.OFF) -> nocloud_net_rules=[] (factory default)
func (p *Provisioner) provisionOff(ctx context.Context) *model.CloudInitResult {
	return &model.CloudInitResult{Mode: model.CloudInitModeOFF, NocloudNetRules: []model.FirewallRule{}}
}

// provisionNet handles NET mode firewall rules for a single VM.
// Uses the pre-allocated server from config (NoCloudURL/NoCloudPort/NoCloudPID),
// or spawns anew via nocloudnetsvc.Spawn if not pre-allocated.
func (p *Provisioner) provisionNet(ctx context.Context) (*model.CloudInitResult, error) {

	// ── Firewall rule creation ──
	if p.firewallTracker == nil {
		slog.Warn("No firewall tracker available, skipping NET mode firewall rules",
			"vm_name", p.config.VMName)
		return &model.CloudInitResult{
			Mode:            model.CloudInitModeNET,
			NocloudURL:      &p.config.NoCloudURL,
			NocloudPort:     p.config.NoCloudPort,
			NocloudPID:      &p.config.NoCloudPID,
			NocloudNetRules: []model.FirewallRule{},
		}, nil
	}

	// Spawn nocloud-net server if not pre-allocated
	if p.config.NoCloudURL == "" {
		port := p.config.NocloudNetPort
		if port == nil || *port == 0 {
			freePort, err := infra.FindFreePort(
				p.config.IPv4Gateway,
				p.config.NocloudPortRangeStart,
				p.config.NocloudPortRangeEnd,
			)
			if err != nil {
				return nil, err
			}
			port = &freePort
		}

		killAfter := p.config.KillAfter
		if killAfter == 0 {
			killAfter = defaultNocloudKillAfter
		}

		result, err := nocloudnetsvc.Spawn(ctx, nocloudnetsvc.Config{
			CloudInitDir: p.config.CloudInitDir,
			Port:         *port,
			Host:         p.config.IPv4Gateway,
			LogFile:      filepath.Join(p.config.VMDir, "nocloud-server.log"),
			KillAfter:    killAfter,
		})
		if err != nil {
			return nil, err
		}

		// Store in config so subsequent calls (e.g. respawn) reuse the same server
		p.config.NoCloudURL = result.URL
		p.config.NoCloudPort = result.Port
		p.config.NoCloudPID = result.PID
	}

	_ = p.firewallTracker.EnsureChain(
		ctx,
		model.FirewallChainMVMNocloudNetIn,
		model.FirewallTableFilter,
		"INPUT",
		0,
	)

	commentTag := fmt.Sprintf("# nocloudnet:%s:%d", p.config.VMName, p.config.NoCloudPort)

	rule := model.FirewallRule{
		TableName:    model.FirewallTableFilter,
		ChainName:    model.FirewallChainMVMNocloudNetIn,
		RuleType:     model.FirewallRuleTypeNocloudNetInput,
		Target:       model.FirewallTargetAccept,
		NetworkID:    p.config.NetworkID,
		Protocol:     model.FirewallProtocolTCP,
		Source:       p.config.GuestIP,
		Destination:  p.config.IPv4Gateway,
		InInterface:  p.config.TapName,
		OutInterface: string(model.FirewallWildcardAnyInterface),
		SPort:        model.FirewallPortAny,
		DPort:        p.config.NoCloudPort,
		NetworkName:  &p.config.NetworkName,
		CommentTag:   &commentTag,
		IsActive:     true,
	}
	if fwResult := p.firewallTracker.EnsureRule(ctx, rule, "nocloud-net"); !fwResult.Success {
		msg := ""
		if fwResult.ErrorMessage != nil {
			msg = *fwResult.ErrorMessage
		}
		return nil, errs.New(errs.CodeCloudInitNetModeFailed,
			fmt.Sprintf("Nocloud-net provisioning failed: %s", msg))
	}

	return &model.CloudInitResult{
		Mode:            model.CloudInitModeNET,
		NocloudURL:      &p.config.NoCloudURL,
		NocloudPort:     p.config.NoCloudPort,
		NocloudPID:      &p.config.NoCloudPID,
		NocloudNetRules: []model.FirewallRule{rule},
	}, nil
}

// provisionISO handles ISO mode — create cloud-init ISO image.
// Matches Python's _provision_iso().
func (p *Provisioner) provisionISO(ctx context.Context) (*model.CloudInitResult, error) {
	// Check for pre-existing custom ISO (Python: if self._config.cloud_init_iso_path is not None)
	if p.config.CloudInitISOPath != nil {
		isoPath := *p.config.CloudInitISOPath
		if _, err := os.Stat(isoPath); os.IsNotExist(err) {
			return nil, errs.New(errs.CodeCloudInitISOModeFailed,
				fmt.Sprintf("Custom cloud-init ISO not found: %s", isoPath),
			)
		}
		return &model.CloudInitResult{
			Mode:            model.CloudInitModeISO,
			ISOPath:         p.config.CloudInitISOPath,
			NocloudNetRules: []model.FirewallRule{},
		}, nil
	}

	// Generate ISO from seed directory
	// Python: except Exception as exc: raise CloudInitIsoModeError(f"Failed to create cloud-init ISO: {exc}") from exc
	isoPath := filepath.Join(p.config.VMDir, p.config.CloudInitISOName)
	if err := p.manager.CreateSeedISO(ctx, p.config.CloudInitDir, isoPath); err != nil {
		return nil, errs.New(errs.CodeCloudInitISOModeFailed,
			fmt.Sprintf("Failed to create cloud-init ISO: %s", err),
		)
	}

	return &model.CloudInitResult{
		Mode:            model.CloudInitModeISO,
		ISOPath:         &isoPath,
		NocloudNetRules: []model.FirewallRule{},
	}, nil
}

// provisionInject handles INJECT mode — config files already written.
// Matches Python's _provision_inject().
func (p *Provisioner) provisionInject(ctx context.Context) (*model.CloudInitResult, error) {
	return &model.CloudInitResult{Mode: model.CloudInitModeINJECT, NocloudNetRules: []model.FirewallRule{}}, nil
}
